<a href="https://colab.research.google.com/github/Artorius81/labs_ml_basov/blob/main/ml_ex_07_12_23.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [15]:
!pip install openpyxl xlrd

In [64]:
import pandas as pd

In [81]:
df = pd.read_excel('/content/data.xlsx')
df

№ п/п      №\nдокумента Дата\nрегистрации              Исх. №\nДата  \
0         1    WSR/1-838/2019        2019-04-09       208-409\n08.04.2019   
1         2    WSR/1-837/2019        2019-04-09        05-338\n09.04.2019   
2         3    WSR/1-835/2019        2019-04-09   И10-14/3137\n09.04.2019   
3         4    WSR/1-834/2019        2019-04-09    Д2/7121-ИС\n03.04.2019   
4         5    WSR/1-833/2019        2019-04-09  15-20/06-911\n08.04.2019   
...     ...               ...               ...                       ...   
4962   4963  WSR/1.-1749/2017        2017-11-01            11\n20.10.2017   
4963   4964  WSR/1.-1751/2017        2017-11-01    ТС-1164/06\n31.10.2017   
4964   4965  WSR/1.-1748/2017        2017-11-01                ЛО-95/06пр   
4965   4966  WSR/1.-1747/2017        2017-11-01    ЛО-93/06пр\n24.10.2017   
4966   4967               NaN               NaT   НА-29/15077\n18.10.2017   

          Адресат                                              Автор  \
0     Уразов Р.Н.                     Антонова С.В. (АО «Интерфакс»)   
1     Уразов Р.Н.  Картошкин С.А. (Министерство просвещения Росси...   
2     Уразов Р.Н.   Цивилев С.Е. (Администрация Кемеровской области)   
3     Иванюк Л.А.  Волков Г.А. (Министерство транспорта Российско...   
4     Уразов Р.Н.  Атанов И.В. (Ставропольский государственный аг...   
...           ...                                                ...   
4962  Уразов Р.Н.  Медведев Д.А. (Аппарат Правительства Российско...   
4963  Уразов Р.Н.  Синюгина Т.Ю. (Министерство образования и наук...   
4964  Уразов Р.Н.  Огородова Л.М. (Министерство образования и нау...   
4965  Уразов Р.Н.  Огородова Л.М. (Министерство образования и нау...   
4966  Уразов Р.Н.  Асаул Н.А. (Министерство транспорта Российской...   

                                    Краткое\nсодержание Примечание  
0     Вх - О возможности использования сетевого изда...        NaN  
1     Вх - Об участии в совещании по вопросу доработ...        NaN  
2     Вх - О командировании Саликовой К. и Глушко Д....        NaN  
3     Вх - Об участии в Чемпионате мира по профессио...        NaN  
4     Вх - Об участии в церемонии открытия Вузовског...        NaN  
...                                                 ...        ...  
4962  Протокол заседания Президиума Совета при Прези...        NaN  
4963  О проведении 21.11.2017 заседания организацион...        NaN  
4964  Протокол заседания рабочей группы приоритетног...        NaN  
4965  Протокол заседания рабочей группы приоритетног...        NaN  
4966  Участие в Чемпионате WorldSkills Hi-Tech Екате...        NaN  

[4967 rows x 8 columns]

**1.1 Предобработка данных и выделение значимых атрибутов**

In [82]:
df.head(15)

№ п/п      №\nдокумента Дата\nрегистрации  \
0       1    WSR/1-838/2019        2019-04-09   
1       2    WSR/1-837/2019        2019-04-09   
2       3    WSR/1-835/2019        2019-04-09   
3       4    WSR/1-834/2019        2019-04-09   
4       5    WSR/1-833/2019        2019-04-09   
5       6    WSR/1-831/2019        2019-04-09   
6       7    WSR/1-830/2019        2019-04-09   
7       8  WSR/1.6-832/2019        2019-04-09   
8       9    WSR/1-836/2019        2019-04-09   
9      10    WSR/1-829/2019        2019-04-08   
10     11  WSR/1.4-828/2019        2019-04-08   
11     12    WSR/1-827/2019        2019-04-08   
12     13    WSR/1-826/2019        2019-04-08   
13     14    WSR/1-825/2019        2019-04-08   
14     15  WSR/1.6-824/2019        2019-04-08   

                        Исх. №\nДата           Адресат  \
0                208-409\n08.04.2019       Уразов Р.Н.   
1                 05-338\n09.04.2019       Уразов Р.Н.   
2            И10-14/3137\n09.04.2019       Уразов Р.Н.   
3             Д2/7121-ИС\n03.04.2019       Иванюк Л.А.   
4           15-20/06-911\n08.04.2019       Уразов Р.Н.   
5             Исх01/0403\n02.04.2019     Миронова С.В.   
6         24-2/10/В-1940\n14.03.2019       Уразов Р.Н.   
7          08-04-21/1408\n08.04.2019     Тымчиков А.Ю.   
8                3571/19\n08.04.2019       Уразов Р.Н.   
9                 05-337\n08.04.2019       Уразов Р.Н.   
10                 140-р\n08.04.2019  Крайчинская С.Б.   
11                380/63\n05.04.2019      Антонов Ю.В.   
12                  40/1\n08.04.2019       Уразов Р.Н.   
13  596\n05.04.2019\n596\n05.04.2019       Уразов Р.Н.   
14             619-45-03\n02.04.2019     Тымчиков А.Ю.   

                                                Автор  \
0                      Антонова С.В. (АО «Интерфакс»)   
1   Картошкин С.А. (Министерство просвещения Росси...   
2    Цивилев С.Е. (Администрация Кемеровской области)   
3   Волков Г.А. (Министерство транспорта Российско...   
4   Атанов И.В. (Ставропольский государственный аг...   
5     Островский А.В. (Губернатор Смоленской области)   
6   Петрова С.В. (Министерство труда и социальной ...   
7   Соппа И.В. (ФГБОУ ВО "Московский политехническ...   
8   Бурганов Р.Т. (Министерство образования и наук...   
9   Черноскутова И.А. (Министерство просвещения Ро...   
10  Ажгиревич А.И. (ОООР "Союз Машиностроителей Ро...   
11  Мишин Д.В. (ФГОБУ ВО "Поволжский государственн...   
12  Цуров М.Т. (ГБПОУ Северо-Кавказский топливно-э...   
13  Кудряшов С.В. (АПОУ Чувашской области "Межреги...   
14  Сагитов С. Т. (ФГБОУ ВО "Башкирский государств...   

                                  Краткое\nсодержание  \
0   Вх - О возможности использования сетевого изда...   
1   Вх - Об участии в совещании по вопросу доработ...   
2   Вх - О командировании Саликовой К. и Глушко Д....   
3   Вх - Об участии в Чемпионате мира по профессио...   
4   Вх - Об участии в церемонии открытия Вузовског...   
5   Вх - Об участии в Чемпионате мира по профессио...   
6   Вх - О направлении протокола "О ходе реализаци...   
7   Вх - О согласовании дат проведения Вузовского ...   
8                      О Национальном чемпионате 2019   
9   Вх - О проведении заседания министров образова...   
10  Вх - Об участии в заочном заседании Совета по ...   
11                         Вх -Об участии в МЧ Казань   
12  Вх - Об отказе рт участия в ОС по компетенции ...   
13  Вх - Об участии в заочном заседании наблюдател...   
14  Вх - О согласовании дат проведения вузовского ...   

                                           Примечание  
0                                                 NaN  
1                                                 NaN  
2                                                 NaN  
3                                                 NaN  
4                                                 NaN  
5                                                 NaN  
6                                                 NaN  
7               

In [83]:
# Здесь мы избавляемся от не нужных для нашей задачи столбцов
df = df.drop(['№ п/п', 'Дата\nрегистрации', 'Исх. №\nДата', '№\nдокумента', 'Примечание'], axis=1)
df

Адресат                                              Автор  \
0     Уразов Р.Н.                     Антонова С.В. (АО «Интерфакс»)   
1     Уразов Р.Н.  Картошкин С.А. (Министерство просвещения Росси...   
2     Уразов Р.Н.   Цивилев С.Е. (Администрация Кемеровской области)   
3     Иванюк Л.А.  Волков Г.А. (Министерство транспорта Российско...   
4     Уразов Р.Н.  Атанов И.В. (Ставропольский государственный аг...   
...           ...                                                ...   
4962  Уразов Р.Н.  Медведев Д.А. (Аппарат Правительства Российско...   
4963  Уразов Р.Н.  Синюгина Т.Ю. (Министерство образования и наук...   
4964  Уразов Р.Н.  Огородова Л.М. (Министерство образования и нау...   
4965  Уразов Р.Н.  Огородова Л.М. (Министерство образования и нау...   
4966  Уразов Р.Н.  Асаул Н.А. (Министерство транспорта Российской...   

                                    Краткое\nсодержание  
0     Вх - О возможности использования сетевого изда...  
1     Вх - Об участии в совещании по вопросу доработ...  
2     Вх - О командировании Саликовой К. и Глушко Д....  
3     Вх - Об участии в Чемпионате мира по профессио...  
4     Вх - Об участии в церемонии открытия Вузовског...  
...                                                 ...  
4962  Протокол заседания Президиума Совета при Прези...  
4963  О проведении 21.11.2017 заседания организацион...  
4964  Протокол заседания рабочей группы приоритетног...  
4965  Протокол заседания рабочей группы приоритетног...  
4966  Участие в Чемпионате WorldSkills Hi-Tech Екате...  

[4967 rows x 3 columns]

In [84]:
# Смотрим на null
df.isnull().sum()

Адресат                0
Автор                  0
Краткое\nсодержание    0
dtype: int64

**Примечание**

Были удалены столбцы '№ п/п', 'Дата\nрегистрации', 'Исх. №\nДата', '№\nдокумента'. Для поставленной задачи они не несут никакой информационной/смысловой нагрузки.

Также столбец 'Примечание' содержит мало записей и также несет мало смысловой нагрузки для нашей задачи. Удаляем.

**1.2 Разбиение сложных атрибутов**

In [85]:
# Импортируем библиотеку для регулярных выражени
import re

In [86]:
# Работаем с "Автор"
df['ФИО'] = df['Автор'].apply(lambda x: re.search(r'^[^()]+', x).group(0))
df['Организация'] = df['Автор'].apply(lambda x: re.search(r'\((.*?)\)', x).group(1))
df = df.drop(columns=['Автор'])
df

Адресат                                Краткое\nсодержание  \
0     Уразов Р.Н.  Вх - О возможности использования сетевого изда...   
1     Уразов Р.Н.  Вх - Об участии в совещании по вопросу доработ...   
2     Уразов Р.Н.  Вх - О командировании Саликовой К. и Глушко Д....   
3     Иванюк Л.А.  Вх - Об участии в Чемпионате мира по профессио...   
4     Уразов Р.Н.  Вх - Об участии в церемонии открытия Вузовског...   
...           ...                                                ...   
4962  Уразов Р.Н.  Протокол заседания Президиума Совета при Прези...   
4963  Уразов Р.Н.  О проведении 21.11.2017 заседания организацион...   
4964  Уразов Р.Н.  Протокол заседания рабочей группы приоритетног...   
4965  Уразов Р.Н.  Протокол заседания рабочей группы приоритетног...   
4966  Уразов Р.Н.  Участие в Чемпионате WorldSkills Hi-Tech Екате...   

                  ФИО                                        Организация  
0      Антонова С.В.                                      АО «Интерфакс»  
1     Картошкин С.А.       Министерство просвещения Российской Федерации  
2       Цивилев С.Е.                   Администрация Кемеровской области  
3        Волков Г.А.        Министерство транспорта Российской Федерации  
4        Атанов И.В.   Ставропольский государственный аграрный универ...  
...               ...                                                ...  
4962   Медведев Д.А.          Аппарат Правительства Российской Федерации  
4963   Синюгина Т.Ю.   Министерство образования и науки РФ. Заместите...  
4964  Огородова Л.М.   Министерство образования и науки Российской Фе...  
4965  Огородова Л.М.   Министерство образования и науки Российской Фе...  
4966      Асаул Н.А.        Министерство транспорта Российской Федерации  

[4967 rows x 4 columns]

In [87]:
# Работаем с "Адресат"
def process_address(column_value):
    names = re.findall(r'[А-ЯЁ][а-яё]+ [А-ЯЁ]\.[А-ЯЁ]\.', column_value)
    return names[0] if names else column_value

df['Адресат'] = df['Адресат'].apply(process_address)
df

Адресат                                Краткое\nсодержание  \
0     Уразов Р.Н.  Вх - О возможности использования сетевого изда...   
1     Уразов Р.Н.  Вх - Об участии в совещании по вопросу доработ...   
2     Уразов Р.Н.  Вх - О командировании Саликовой К. и Глушко Д....   
3     Иванюк Л.А.  Вх - Об участии в Чемпионате мира по профессио...   
4     Уразов Р.Н.  Вх - Об участии в церемонии открытия Вузовског...   
...           ...                                                ...   
4962  Уразов Р.Н.  Протокол заседания Президиума Совета при Прези...   
4963  Уразов Р.Н.  О проведении 21.11.2017 заседания организацион...   
4964  Уразов Р.Н.  Протокол заседания рабочей группы приоритетног...   
4965  Уразов Р.Н.  Протокол заседания рабочей группы приоритетног...   
4966  Уразов Р.Н.  Участие в Чемпионате WorldSkills Hi-Tech Екате...   

                  ФИО                                        Организация  
0      Антонова С.В.                                      АО «Интерфакс»  
1     Картошкин С.А.       Министерство просвещения Российской Федерации  
2       Цивилев С.Е.                   Администрация Кемеровской области  
3        Волков Г.А.        Министерство транспорта Российской Федерации  
4        Атанов И.В.   Ставропольский государственный аграрный универ...  
...               ...                                                ...  
4962   Медведев Д.А.          Аппарат Правительства Российской Федерации  
4963   Синюгина Т.Ю.   Министерство образования и науки РФ. Заместите...  
4964  Огородова Л.М.   Министерство образования и науки Российской Фе...  
4965  Огородова Л.М.   Министерство образования и науки Российской Фе...  
4966      Асаул Н.А.        Министерство транспорта Российской Федерации  

[4967 rows x 4 columns]

**Примечание**

С помощью регулярных выражений столбец "Автор" был разбит на два ("ФИО" и "Организация"). Также в "Адресат" в некоторых строках встречается списко ФИО для отправки. Для удобства сделали, чтобы в таких строках записывало только первый элемент списка.

**1.3 Дополнение недостающими данными**



**Примечание**

По заданию:

*Такая же ситуация с адресатом, система, определяющая направление документа по фамилии будет являться частным случаем решения, и в случае кадровых перестановок не сможет функционировать.*

Алгоритм решения:



1.   Создаем словарь с названиями тех организация, из которых письма идут сразу к Генеральному Директору (смотрим Алгоритм.pdf). Также смотрим по вопросам исполнения.
2.   



**1.4 Визуализация зависимостей данных**